### ISSUE TITLE
No Access anymore

### ISSUE THREAD

**Issue by xerion3800** [Hacksore/bluelinky#101]
Since today 21-02-2021 i didn't get any access anymore. From a Hyundai guy I heard that they want to change the authentication process. I'am user of Europe maybe it is only an issue in Europe.

**Comment by tomtom5152** [comment]
I concur with this, from my own digging it appears as if the notification registration step now requires an application ID as well as a service ID. I am attempting to reverse engineer it a bit further with mitmproxy.

**Comment by neoPix** [comment]
I'm working with Kuvork (a fork for UVO) and it's the same.
Debugging it : the firsts steps of the auth are ok (for now),
then the `/api/v1/spa/notifications/register` fails with this error : 

```
  statusCode: 401,
  statusMessage: 'Unauthorized',
  body: {
    retCode: 'F',
    resCode: '4017',
    resMsg: 'Invalid request value - Invalid Application Id.',
    msgId: 'ea77e3f4-3f25-4211-9193-1cef7f7bc525'
  }
```

**Comment by tomtom5152** [comment]
From what I've been able to determine by decompiling the APK and network sniffing, there is now a `Stamp` header which so far I've been able to determine is formed by (psudo) `base64_encode(AES_encrypt(<app_id> + ':' + <unix_timestamp_ms>))`. 

I am currently working to determine what the key used for this encryption operation is, but its pretty clear this is an obvious step to stop 3rd party API access.

**Comment by PierreLevres** [comment]
> From what I've been able to determine by decompiling the APK and network sniffing, there is now a `Stamp` header which so far I've been able to determine is formed by (psudo) `base64_encode(AES_encrypt(<app_id> + ':' + <unix_timestamp_ms>))`.
> 
Yes, I noticed that one too. Are you subscribed to the discord server?

**Comment by tomtom5152** [comment]
I am not, any chance of an invite please?

**Comment by PierreLevres** [comment]
> I am not, any chance of an invite please?
> […](#)

It is in the readme.md https://discord.com/invite/HwnG8sY

**Comment by tomtom5152** [comment]
A little more digging this morning shows that the call to register for GCM is now failing as well, looks as if they have changed their appID there as well.

**Comment by fredrikedling** [comment]
> A little more digging this morning shows that the call to register for GCM is now failing as well, looks as if they have changed their appID there as well.

For information got the information from the Bluelink support in Europe last week that a completely new app will be launched around May so maybe the change in appID could be related to that.

**Comment by dlysnes** [comment]
Confirmed same issue( EU). I'm using bluelinky to wake up my Hyundai Kona from hibernation from a raspberry pi as there's a bug in the onboard charger software. It will not start charging when power is turned on otherwise, and Hyundai  isn't putting much effort in fixing this..
Any more digging done?

**Comment by dansloane** [comment]
For reference, Ohme have identified API changes that break their chargers for both Kia and Hyundai.   They've emailed me to say they're working on a fix, however it confirms the issues seen here.

**Comment by mpucherSmIT** [comment]
Any updates here?

**Comment by Hacksore** [comment]
@mpucherSmIT I think the plan is to generate a stamp list and include them in the library for use at runtime.

@neoPix is that going to be the plan for #105? Can omit the tools from this repo and leave the generation to https://github.com/Hacksore/hyundai-kia-stamp/

**Comment by neoPix** [comment]
@PierreLevres as an implementation of [Kuvork@6.2.0-alpha.5](https://www.npmjs.com/package/kuvork/v/6.2.0-alpha.5) which is working well based on what you are suggesting. I'll adjust the PR to use the docker call at build time.

**Comment by PierreLevres** [comment]
Not entirely correct. The dockers leaves a text file, with a <stamp> on each line. 
The code of #105 needs a .ts file with a structure like. 
export default ["<stamp1>","<stamp2>",..."<stampn>"]

Either the docker, or the code of #105 needs to be rewritten a bit. Since opening, reading, and closing a file on runtime for each call is rather expensive on disk and memory, I would suggest only to open, read a stamp  and close the file once at login in the session controller.
For my Kuvork implementation in the short run I added just 5 stamps by hand, not the 100.000 :). Since I am not a JS developer and don't wanna drift away from BlueLinky, I don't wanna touch that code just yet.

**Comment by neoPix** [comment]
I just updated #105 to match this discussion :wink:.

**Comment by Hacksore** [comment]
Should be fixed in `v6.0.1` via #105
